src/features/tokens/CosmNativeTokenAdapter.ts

```diff
@@ -0,0 +1,135 @@
+      throw new Error('Missing properties for CosmNativeTokenAdapter');
+    super(chainName, multiProvider, addresses);
+  }
+
+  async getBalance(address: string): Promise<string> {
+    const provider = await this.getProvider();
+    const coin = await provider.getBalance(address, this.properties.ibcDenom);
+    return coin.amount;
+  }
+
+  getMetadata(): Promise<MinimalTokenMetadata> {
+    throw new Error('Metadata not available to native tokens');
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  populateApproveTx(_transferParams: TransferParams): unknown {
+    throw new Error('Approve not required for native tokens');
+  }
+
+  // TODO cosmos consider refactoring this into a new method b.c. the other adapter
+  // transfer methods are for local chain transfers, not remote ones
+  async populateTransferTx(
+    transferParams: TransferParams,
+    memo = '',
+  ): Promise<MsgTransferEncodeObject> {
+    if (!transferParams.fromAccountOwner)
+      throw new Error('fromAccountOwner is required for ibc transfers');
+
+    const value = {
+      sourcePort: this.properties.sourcePort,
+      sourceChannel: this.properties.sourceChannel,
+      token: {
+        denom: this.properties.ibcDenom,
+        amount: transferParams.weiAmountOrId.toString(),
+      },
+      sender: transferParams.fromAccountOwner,
+      receiver: transferParams.recipient,
+      // Represented as nano-seconds
+      timeoutTimestamp: Long.fromNumber(
+        new Date().getTime() + COSMOS_IBC_TRANSFER_TIMEOUT,
+      ).multiply(1_000_000),
+      memo,
+    };
+    return {
+      typeUrl: '/ibc.applications.transfer.v1.MsgTransfer',
+      value,
+    };
+  }
+}
+
+export class CosmNativeToHypTokenAdapter extends CosmNativeTokenAdapter implements ITokenAdapter {
+  constructor(
+    public readonly chainName: ChainName,
+    public readonly multiProvider: MultiProtocolProvider,
+    public readonly addresses: {
+      intermediateRouterAddress: Address;
+      destinationRouterAddress: Address;
+    },
+    public readonly properties: CosmNativeTokenAdapter['properties'] & {
+      derivedIbcDenom: string;
+      intermediateChainName: ChainName;
+      destinationDomainId: DomainId;
+    },
+  ) {
+    super(chainName, multiProvider, addresses, properties);
+  }
+
+  async populateTransferTx(transferParams: TransferParams): Promise<MsgTransferEncodeObject> {
+    const cwAdapter = new CwHypCollateralAdapter(
+      this.properties.intermediateChainName,
+      this.multiProvider,
+      {
+        token: this.properties.derivedIbcDenom,
+        warpRouter: this.addresses.intermediateRouterAddress,
+      },
+      this.properties.ibcDenom,
+    );
+    const transfer = await cwAdapter.populateTransferRemoteTx({
+      ...transferParams,
+      destination: this.properties.destinationDomainId,
+      // TODO cosmos quote real interchain gas payment
+      txValue: '1',
```

Comment:
No action needed here, but this is probably where we would have the logic that charges some amount of the token that is being transferred as IGP payment on the forwarding chain?

Reply:
Yes this is a possible place for that